### PR TITLE
feat(api-client): Create controller for Project module

### LINKED ISSUE #347: API-CLIENT: Create controllers to make request on behalf of the client apps
## Description

The [`api-client`](https://github.com/keyshade-xyz/keyshade/tree/develop/packages/api-client) package contains a set of controllers to make calls to the API. We would use these controllers from the CLI and the platform to perform all the API related operations. As a result, we would like to have all the client requests in this `api-client` package.

Currently, we just have the controller for `Environment` module, but we would like to add all the other ones aswell. Here's the list:

- Workspace
- User
- Project
- Variable
- Secret
- Integration
- User
- Workspace Role
- Workspace Membership
- Event

## Solution

This is a parent issue, and all other issues linked to this would follow the same pattern. Just that the scope of implementation will vary. Here are the general pointers that should be implemented:

- Create a `controller` under [`src/controllers`](https://github.com/keyshade-xyz/keyshade/tree/develop/packages/api-client/src/controllers/) and name it as `<module-name>.ts`.
- Implement all the API calls from within this controller. Refer to [this](https://github.com/keyshade-xyz/keyshade/blob/develop/packages/api-client/src/controllers/environment/environment.ts)
- Add the request and response types under `src/types/<module-name>.types.d.ts`. Refer to [this](https://github.com/keyshade-xyz/keyshade/blob/develop/packages/api-client/src/types/environment.types.d.ts)
- Add relevant tests under `tests/<module-name>.spec.ts`. Refer to [this](https://github.com/keyshade-xyz/keyshade/blob/develop/packages/api-client/tests/environment.spec.ts) 

## Objectives

- #348
- #349
- #350
- #351
- #352
- #353
- #354
- #355
- #356
- #446

### LINKED ISSUE #350: API-CLIENT: Create controller for Project module
## Description 

Refer to #347 

API Controller: https://github.com/keyshade-xyz/keyshade/blob/develop/apps/api/src/project/controller/project.controller.ts

### PR DESCRIPTION
**Pull request by vr-varad**
## Description

Api client controller for project

Fixes #350 

## Dependencies

_Mention any dependencies/packages used_

## Future Improvements

_Mention any improvements to be done in future related to any file/feature_

## Mentions

_Mention and tag the people_

## Screenshots of relevant screens

_Add screenshots of relevant screens_

## Developer's checklist

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-check on my work

**If changes are made in the code:**

- [x] I have followed the [coding guidelines](https://google.github.io/styleguide/jsguide.html)
- [x] My changes in code generate no new warnings
- [ ] My changes are breaking another fix/feature of the project
- [ ] I have added test cases to show that my feature works
- [ ] I have added relevant screenshots in my PR
- [ ] There are no UI/UX issues

### Documentation Update

- [ ] This PR requires an update to the documentation at [docs.keyshade.xyz](https://docs.keyshade.xyz)
- [ ] I have made the necessary updates to the documentation, or no documentation changes are required.

### CODE REVIEW

**Review by rajdip-b — COMMENTED** [review]
The structure looks good.

File: packages/api-client/src/controllers/project/project.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+      `/api/project/${request.projectId}/forks`,
+      headers
+    )
+  }
+
+  static async getAllProjects(
+    request: GetAllProjectsRequest,
+    headers: any
+  ): Promise<any> {
+    return this.apiClient.get(`/api/project/all/${request.projectId}`, headers)
```
Comment (rajdip-b) [inline packages/api-client/src/controllers/project/project.ts:87]: We need to set the pagination properly in here. Check how it works in environment.

File: packages/api-client/src/controllers/project/project.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    headers: any
+  ): Promise<any> {
+    return this.apiClient.post(
+      `/api/project/${request.projectId}/fork`,
+      request,
+      headers
+    )
+  }
+
+  static async syncFork() {}
```
Comment (rajdip-b) [inline packages/api-client/src/controllers/project/project.ts:63]: Missing implementation.

File: packages/api-client/src/controllers/project/project.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    headers: any
+  ): Promise<any> {
+    return this.apiClient.put(
+      `/api/project/${request.projectId}/unlink-fork`,
+      request,
+      headers
+    )
+  }
+
+  static async getForks(request: GetForkRequest, headers: any): Promise<any> {
```
Comment (rajdip-b) [inline packages/api-client/src/controllers/project/project.ts:76]: Missing pagination

File: packages/api-client/src/types/project.types.d.ts
Comment (rajdip-b) [inline packages/api-client/src/types/project.types.d.ts:1]: Type definitions missing

File: packages/api-client/src/controllers/project/project.ts [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    return this.apiClient.post(
+      `/api/project/${request.projectId}/fork`,
+      request,
+      headers
+    )
+  }
+
+  static async syncFork() {}
+
+  static async unlinkFork(
```
Comment (rajdip-b) [inline packages/api-client/src/controllers/project/project.ts:65]: I've updated the path for this to `DELETE /api/project/{projectId}/fork`

**Review by rajdip-b — CHANGES_REQUESTED** [review]
You would also need to include tests

File: packages/api-client/src/controllers/project/project.ts (hunk 2 above)
Comment (vr-varad) [reply packages/api-client/src/controllers/project/project.ts:63]: i couldn't find api for it in postman @rajdip-b

Comment (rajdip-b) [reply packages/api-client/src/controllers/project/project.ts:63]: Unfortunately that one is outdated aswell :/ Until the next release on monday. You would need to infer it from the controller in project module.

File: packages/api-client/src/controllers/project/project.ts [hunk 5]
```diff
@@ -64,26 +75,37 @@ export default class ProjectController {
...
+  static async getForks(
+    request: GetForkRequest,
+    headers: Record<string, string>
+  ): Promise<[GetForkResponse]> {
+    let url = `/api/project/${request.projectId}/forks`
+    request.page && (url += `page=${request.page}&`)
+    request.limit && (url += `limit=${request.limit}&`)
+    request.sort && (url += `sort=${request.sort}&`)
+    request.order && (url += `order=${request.order}&`)
+    request.search && (url += `search=${request.search}&`)
```
Comment (rajdip-b) [inline packages/api-client/src/controllers/project/project.ts:95]: image

We have just these few fields for get forks

**Review by rajdip-b — CHANGES_REQUESTED** [review]

**Comment by vr-varad** [comment]
@rajdip-b wrote the tests but getting this error any idea why?
 RUNS   api  tests/project.spec.ts
/home/varad/Desktop/keyshade/packages/api-client:
 ERR_PNPM_RECURSIVE_RUN_FIRST_FAIL  api-client@1.0.0 test: `BACKEND_URL=http://localhost:4200 jest`
Exit status 1
 WARN   Local package.json exists, but node_modules missing, did you mean to install?
 ELIFECYCLE  Command failed with exit code 1.
husky - pre-commit script failed (code 1)

**Comment by rajdip-b** [comment]
I've pushed an update that fixes this error.

**Comment by vr-varad** [comment]
@rajdip-b the tests are running successfully

**Comment by rajdip-b** [comment]
cool! I'll test them and approve it

**Review by rajdip-b — APPROVED** [review]

**Pull request merged**